source/agora/common/DNS.d

```diff
@@ -1107,6 +1108,40 @@ public struct Domain
         return result;
     }
 
+    /// Tuple for service and protocol labels
+    alias ServiceProtoTuple = Tuple!(const(char[]), "service", const(char[]), "protocol");
+
+    /***************************************************************************
+
+        Get service and protocol labels
+
+        Returns:
+          A tuple containing both service and protocol labels, empty
+          `ServiceProtoTuple` is returned when one of them is missing.
+
+    ***************************************************************************/
+
+    public ServiceProtoTuple serviceAndProtocolLabels () const nothrow @safe
+    {
+        if (this.value[0] != '_')
+            return ServiceProtoTuple.init;
+
+        try
+        {
+            auto splitted = this.value.splitter('.');
+            const service = splitted.front();
+            splitted.popFront();
+            const proto = splitted.front();
```

Comment:
That's not correct, you need to check if the range is `empty` at some point.

Reply:
Since we always have a trailing `.` we always have an empty label; but this still misses check for empty label, added.

Reply:
Try "." and it still `assert`s.